Library/Homebrew/utils/path.rb

```diff
@@ -10,5 +10,53 @@ def self.child_of?(parent, child)
       child_pathname.ascend { |p| return true if p == parent_pathname }
       false
     end
+
+    sig {
+      params(
+        path: Pathname,
+        allowed_paths: T::Array[String],
+        package_type: String,
+        create_command: String
+      ).returns(Symbol)
+    }
+    def self.validate_package_path(path, allowed_paths, package_type, create_command)
+      return :allow unless Homebrew::EnvConfig.forbid_packages_from_paths?
+
+      path_realpath = path.realpath.to_s
+      path_string = path.to_s
+
+      # Check if this is a Ruby file and not in allowed paths
+      if (path_realpath.end_with?(".rb") || path_string.end_with?(".rb")) &&
+         !allowed_paths.any? { |allowed| path_realpath.start_with?(allowed) } &&
+         !allowed_paths.any? { |allowed| path_string.start_with?(allowed) }
+
+        # Check for conditions that should cause rejection
+        if path_string.include?("./") || path_string.end_with?(".rb") || path_string.count("/") != 2
+          :reject_with_error
+        elsif path_string.count("/") == 2
+          # Looks like a tap, let's quietly return but not error.
+          :reject_silently
+        else
+          :allow
+        end
+      else
+        :allow
+      end
+    end
+
+    sig { params(path_string: String, path_realpath: String, package_type: String, create_command: String).returns(String) }
+    def self.package_path_error_message(path_string, path_realpath, package_type, create_command)
+      singular_type = package_type == "formulae" ? "formula" : "cask"
+      
+      <<~WARNING
+        Homebrew requires #{package_type} to be in a tap, rejecting:
+          #{path_string} (#{path_realpath})
```

Comment:
```suggestion
    sig { params(path_string: Pathname, package_type: String, create_command: String).returns(String) }
    def self.package_path_error_message(path, package_type, create_command)
      singular_type = package_type == "formulae" ? "formula" : "cask"
      
      <<~WARNING
        Homebrew requires #{package_type} to be in a tap, rejecting:
          #{path} (#{path.realpath})
```

We can handle the path to string conversion in this method to make calling this simpler